### DISCUSSION TITLE
docs/guides/polymorphism

### DISCUSSION THREAD

**Discussion by giscus** [zenstackhq/zenstack-docs#370] [Comments]
# docs/guides/polymorphism

Polymorphic Relations

https://zenstack.dev/docs/guides/polymorphism

**Comment by bbozzay** [comment]
A note for others reading this: you probably don't want to ever have an enhanced prisma client with only the delegate kind specified since this would disable the access policy (and the other features as well).

**Comment by PodkopovP** [comment]
Why would count, aggregate, and groupBy not allow RLS on the base fields?

I might be misunderstanding, but I thought extending a base model is essentially just using it as a template to write less boilerplate?

**Reply by ymc9** [reply]
Extending from a "delegate" base is more than boilerplate. Under the hood it implements the delegate types (aka multi-table-inheritance) pattern for you as described here: https://zenstack.dev/blog/polymorphism#3-delegated-types

Since the inheritance is really backed by relations, it inherits Prisma limitations that for aggregations you can't traverse relation fields. There're several Prisma issues related to it.

**Comment by PodkopovP** [comment]
Ok thank you for explaining - so with that in mind, is there a way to create boilerplate templates that can be reused along with RLS etc?

If not, it would be super useful - I’ve got a common pattern across pretty much every table (Id, created on, updated on, deleted on), with the deleted on omitted and deleted rows have no access at all.

Further to that, a common user table is extended with relations to the user, visible to that user and hidden to blocked users.

Every single table follows these patterns, and uses values from them to further enable/restrict access.

**Reply by ymc9** [reply]
Maybe what you're looking for is this? https://zenstack.dev/docs/guides/multiple-schema#abstract-model-inheritance

**Comment by eleazareramos** [comment]
RE: This current limitation: __Inheriting from multiple `@delegate` models is not supported yet.__

Is this on the current roadmap for V3? :)

**Reply by ymc9** [reply]
Not yet, I feel multi-inheritance can potentially introduce a bunch of new complexities. Do you have a use case for it?

**Reply by eleazareramos** [reply]
Just thinking about general polymorphism principles, which allow for multiple levels for interface definition:

For example:
- Animal
  - Dog
    - Terrier
    - Hound
  - Cat
    - Siamese
    - Persian
    
Or following the current examples from the docs...

- Content
  - Post
    - Article
    - Note
  - Video
 
Where `Article` and `Note` inherit properties from `Post` and `Content` but have their own specific properties.

I can see how that can be complex due to the infinitely recursive nature... but can be impactful for true polymorphism.

Anyway, thanks for the response / update! Please consider this an upvote if that's still on the roadmap! :)

**Comment by IanvanZyl636** [comment]
A question if you have say Media and Image extends Media and in Image I have a parent child setup meaning one image is the parent of other generated images how can I include the children property in a prisma query when I access the base like the following model. prisma.media.findFirst({where:{id}, include:{children:true}} children only exists on the extended Image model. The only work around I have is to query the Image model again to get the children which is an extra DB call that I want to avoid.
